src/mlmc/quantity_concept.py

```
@@ -1,109 +1,313 @@
 
-    def _samples(self, i_chunk):
+    @cached(custom_key_maker=get_cache_key)
+    def samples(self, level_id, i_chunk):
         """
         Yields list of sample chunks for individual levels.
-        Possibly calls underlaying quantities.
-        Try to write into the chunk array.
-        TODO: possibly pass down the full table of composed quantities and store comuted temporaries directly into the composed table
-
-        TODO: possible problem when more quantities dependes on a single one, this probabely leads to
-        repetitive evaluation of the shared quantity
-        This prevents some coies in concatenation.
+        Possibly calls underlying quantities.
+        :param level_id: int
+        :param i_chunk: int
+        :return: np.ndarray
         """
-        chunks_quantity_level = [q._samples(i_chunk) for  q in self._input_quantities]
+        chunks_quantity_level = [q.samples(level_id, i_chunk) for q in self._input_quantities]
+
         is_valid = (ch is not None for ch in chunks_quantity_level)
         if any(is_valid):
             assert (all(is_valid))
-            chunks_level_quantity = zip(*chunks_quantity_level)
-            return [self.operation(*one_level_chunks) for one_level_chunks in chunks_level_quantity]
+            # Operation not set return first quantity samples - used in make_root_quantity
+            if self._operation is None:
+                return chunks_quantity_level[0]
+
+            try:  # numpy function does not have __code__
+                # We need to pass level id and chunk id to select operation
+                if all(par in self._operation.__code__.co_varnames for par in ['level_id', 'i_chunk']):
+                    return self._operation(*chunks_quantity_level, level_id=level_id, i_chunk=i_chunk)
+            except:
+                pass
+
+            return self._operation(*chunks_quantity_level)
         else:
             return None
```

Should we raise an exception instead of returning None? there should be no None among the input chunks.